File: 132Pattern6.py
class Solution:
    def find132pattern(self, nums) -> bool:
        stack = []
        for num in nums:
            if not stack or stack[-1][0] > num:
                stack.append([num, num])
            elif stack[-1][0] < num:
                last = stack.pop()
                if num < last[1]:
                    return True
                else:
                    last[1] = num
                    while stack and stack[-1][1] <= num:
                        stack.pop()
                    if stack and num > stack[-1][0]:
                        return True
                    stack.append(last)
        return False


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.find132pattern(nums))

File: 132Pattern6.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<stack>

using namespace std;

class Solution {
public:
    bool find132pattern(vector<int>& nums) {
    	stack<pair<int, int> > s;
    	for(int num : nums) {
    		if(s.empty() || num < s.top().first) s.push(make_pair(num, num));
    		else if(num > s.top().first){
    			auto last = s.top();
    			s.pop();
    			if(num < last.second) return true;
    			else {
    				last.second = num;
    				while(!s.empty() && num >= s.top().second) s.pop();
    				if(!s.empty() && s.top().first < num ) return true;
    				s.push(last);
				}
			}
		}
		return false;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->find132pattern(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}
